Experimental direct role execution

matburt opened the pull request:
<!--- changelog-entry
# Fill in 'msg' below to have an entry automatically added to the next release changelog.
# Leaving 'msg' blank will not generate a changelog entry for this PR.
# Please ensure this is a simple (and readable) one-line string.
---
msg: "Adds support for direct role execution defined by properties on the job template"
-->

##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
See this old issue: https://github.com/ansible/awx/issues/3780

This is work towards adding Direct Role execution capabilities into AWX. Since we're moving towards Execution Environments as the mode of execution and Collections as one of the primary delivery pipelines adding support for this gives another vector for delivering automation directly to Controller itself.

Having direct role execution also enables more targeted use of Role Arg Spec as a capability within Ansible core itself, potentially being able to inform Survey semantics. This also leads towards being able to select playbooks from Collections as a source of automation.

The functionality provided here may also lead to directly publishing content from GalaxyNG and Dev Tools where those systems have a deep understanding of the source and provenance of the automation content but don't have the capability to create and publish a source control repo. Imagine a scenario where a developer is testing a role or a playbook and has the ability to push it directly into AWX organization specifically confined for automation testing.

This change also encourages users and organizations to standardize on automation content publishing without needing to expose source control semantics to the automation system itself.

The way this PR is implemented is that it removes the hard requirement that a project and playbook be defined and instead allows either a Project/Playbook to be used OR the FQCN+Role be specified. This will trigger different behavior in Ansible Runner.

This PR is missing some things that would make it generally more useful for an integration:

- [x] Tests
- [ ] Documentation
- [ ] Collection and Role indexing from the Execution Environment itself. This is currently done in a partial way from the Job runtime and should be changed so that the information is also stored adjacent to the EE.
- [ ] UI/UX -- The UI is deeply confused and concerned that there's no project or playbook assigned 
- [ ] We can probably take care of https://github.com/ansible/awx/issues/10409 at the same time, as right now there's just a placeholder value for the SCM revision.
- [ ] Direct Playbook Execution from a Collection
(More that I'm sure will come up as part of discussion)

There are some adjacent things I'm using as part of the validation while working on this, namely:

- https://github.com/ansible/ansible-tower-samples/pull/26 is a PR on the demo git repo that allows generating a collection from its content
- The EE definition that makes use of that change to produce an EE containing that collections: https://gist.github.com/matburt/94c5410a4ba7da20c9cf09ce07091c73
- The EE (quay.io/matburt/ansible-sample-ee:latest) that I've published to my quay account that is the manifestation of both of those things: https://quay.io/repository/matburt/ansible-sample-ee?tab=tags



<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API
 - UI
 - Collection

##### AWX VERSION
<!--- Paste verbatim output from `make VERSION` between quotes below -->
```
HEAD
```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->
![directro8](https://user-images.githubusercontent.com/89544/169078279-c2580910-d068-429c-a5f2-2cd7e9a33cb7.gif)

<!--- Paste verbatim command output below, e.g. before and after your change -->
```

```

matburt committed:
Experimental direct role execution

* Adds a role fqcn path parameter to a job template definition
* Changes the conditional logic for job launch to consider that parameter as an alternative to project/playbook

matburt commented on awx/api/serializers.py:
```diff
@@ -2778,6 +2778,7 @@ class Meta:
             'inventory',
             'project',
             'playbook',
+            'fqcn_role',
```
Not sure how I feel about the name here... "role" is mostly already defined for the rbac system.

matburt commented on awx/api/serializers.py:
```diff
@@ -2828,25 +2829,27 @@ def to_representation(self, obj):
 
     def validate(self, attrs):
-        if 'project' in self.fields and 'playbook' in self.fields:
+        if 'project' in self.fields and 'playbook' in self.fields and 'fqcn_role' in self.fields:
```
I feel like a lot of these conditionals should be reworked, but for a first stab at it... these work.

matburt commented on awx/api/serializers.py:
```diff
@@ -2828,25 +2829,27 @@ def to_representation(self, obj):
-                raise serializers.ValidationError({'scm_branch': _('Project does not allow overriding branch.')})
-            if ask_scm_branch_on_launch and not project.allow_override:
-                raise serializers.ValidationError({'ask_scm_branch_on_launch': _('Project does not allow overriding branch.')})
+            if fqcn_role in (None, ''):
```
The potential duality of testing these makes these conditional checks really uncomfortable.

matburt commented on awx/api/serializers.py:
```diff
@@ -2828,25 +2829,27 @@ def to_representation(self, obj):
-                raise serializers.ValidationError({'ask_scm_branch_on_launch': _('Project does not allow overriding branch.')})
+            if fqcn_role in (None, ''):
+                if not project:
+                    raise serializers.ValidationError({'project': _('This field is required.')})
```
We don't ... that I'm aware ... have a way to represent something like this in an error "Either do this OR do this other thing"

matburt commented on awx/main/models/jobs.py:
```diff
@@ -291,14 +297,13 @@ def validation_errors(self):
     @property
     def resources_needed_to_start(self):
-        return [fd for fd in ['project', 'inventory'] if not getattr(self, '{}_id'.format(fd))]
+        return [fd for fd in ['inventory'] if not getattr(self, '{}_id'.format(fd))]
```
This is a mostly unacceptable change, but it gets us past the check itself. There's a complex layer of validation on these properties and it feels like we need another representation. At least at this level.

matburt commented on awx/main/tasks/jobs.py:
```diff
@@ -10,6 +10,7 @@
 from pathlib import Path
 import shutil
 import stat
+from xxlimited import foo
```
Not sure where this came from, I'll remove it

matburt committed:
Remove spurious import

AlanCoding commented on awx/api/serializers.py:
```diff
@@ -2970,13 +2973,14 @@ def get_field_from_model_or_attrs(fd):
-        if project is None:
-            raise serializers.ValidationError({'project': _("Job Templates must have a project assigned.")})
+        if project is None and fqcn_role == '':
+            raise serializers.ValidationError({'project': _("Job Templates must have a project or role assigned.")})
```
For validation - should `execution_environment` be required if `fqcn_role` is set?

Otherwise you would be using the project / default EE. While yes, that might conceptually work, in a way, it doesn't make sense. Providing the FQCN path is creating a hard dependency on the execution environment. Ultimately (even if not now), the user experience for doing this should start with the EE.

matburt replied:
Possibly... I think my question would be, is there a scenario where the user overrides the default EE and thus expects that default EE to be capable of role execution without the EE being supplied to the Job Template?

matburt replied:
Collecting details about whether or not a collection and role is available in the selected EE is probably worth persuing as an indication to the user that this job might not run with the currently selected EE. It is entirely possible that the EE could materialize the collection and role at runtime, though.

AlanCoding commented on awx/main/tasks/jobs.py:
```diff
@@ -845,48 +849,52 @@ def pre_run_hook(self, job, private_data_dir):
+                sync_needs.extend(['install_roles', 'install_collections'])
+        else:
+            sync_needs = []
+            job_revision = "collection"
```
We're here if `job.fqcn_role` is not blank...

Would this allow running from a project's collection requirements? In that case you'd have both a project and a role path. This stuff could get weird.

matburt replied:
This is why I make this an either/or thing.... maybe I could be persuaded against that but I'm drawing a line here that says: No project + FQCN. The alternative is to have this be entirely vs a selected playbook but that removes some of the utility of just being able to selected an EE sans project.

AlanCoding commented:
I had imagined direct role execution as more of a special case of ad hoc commands, as opposed to job templates. Although I'm still thinking through that, if the underlying command is `ansible`, I think it should be an ad hoc command. If the underlying command is `ansible-playbook`, what you're doing here would be preferable.

matburt commented:
Honestly, I'd like to get away from the way we do ad-hoc as a concept. It's a dark pattern and hidden behind the inventory interface unnecessarily.

matburt committed:
Ensure role parameter exists before check

This works around an issue where some Job Template types don't have this parameter

AlanCoding commented:
https://docs.ansible.com/ansible/latest/user_guide/intro_adhoc.html

https://ansible-runner.readthedocs.io/en/stable/standalone/?highlight=role#running-roles-directly

Well, it's true that direct role execution is not a canonical Ansible core construct, and I see that ansible-runner writes a playbook on the fly (sorry this was not obvious to me at the start). Content being:

```python
       play = [{'hosts': vargs.get('hosts') if vargs.get('hosts') is not None else "all",
                 'gather_facts': not vargs.get('role_skip_facts'),
                 'roles': [role]}]
```

These implementation details have a lot of impact on the user experience.

matburt committed:
Adding role execution unit test

matburt committed:
Adding support for direct role execution in collection

AlanCoding commented:
I'm going to close this, as I would like to revive some still-workable related issues, like the ability to run playbooks from a collection.
